File: Sources/LRSession.m
//
//  LRSession.m
//  LoginRadiusSDK
//
//  Copyright © 2016 LoginRadius Inc. All rights reserved.
//
//

#import "LRSession.h"
#import "LRMutableDictionary.h"
#import "LoginRadiusSDK.h"
#import "SimpleKeychain.h"
#import "LoginRadius.h"

@interface LRSession()
@end

@implementation LRSession

+ (instancetype)sharedInstance {
	static dispatch_once_t onceToken;
	static LRSession *instance;
	dispatch_once(&onceToken, ^{
		instance = [LRSession instance];
	});

	return instance;
}

+ (instancetype)instance {

    return [[LRSession alloc] init];
}

- (nullable NSString*) accessToken {
    
    NSString *token = nil;
    
    if([[LoginRadiusSDK sharedInstance] useKeychain])
    {
        //retrive keychain
        //KeychainAccessVersionNumber
        NSString *appIdentifierPrefix = [self bundleSeedID];
        
        NSString *groupKey = [NSString stringWithFormat:@"%@.%@",appIdentifierPrefix,[LoginRadiusSDK siteName]];

        A0SimpleKeychain *keychain = [A0SimpleKeychain keychainWithService:[LoginRadiusSDK siteName] accessGroup:groupKey];
        
        token = [keychain stringForKey:@"lrAccessToken"];
    }else
    {
        NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
        token = [defaults valueForKey:@"lrAccessToken"];
        
    }
    
    return token;
}

- (nullable NSDictionary*) userProfile {
    
    NSDictionary *profile = nil;
    
    if([[LoginRadiusSDK sharedInstance] useKeychain])
    {
        //retrive keychain
        NSString *appIdentifierPrefix = [self bundleSeedID];
        
        NSString *groupKey = [NSString stringWithFormat:@"%@.%@",appIdentifierPrefix,[LoginRadiusSDK siteName]];

        A0SimpleKeychain *keychain = [A0SimpleKeychain keychainWithService:[LoginRadiusSDK siteName] accessGroup:groupKey];
        
        NSData *profileData = [keychain dataForKey:@"lrUserProfile"];
        NSDictionary *profileDict = (NSDictionary*) [NSKeyedUnarchiver unarchiveObjectWithData:profileData];

        profile = profileDict;
    }else
    {
        NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
        profile = [defaults valueForKey:@"lrUserProfile"];
        
    }
    
    return profile;
}

- (instancetype)initWithAccessToken:(NSString*_Nonnull)token userProfile:(NSDictionary*_Nonnull)userProfile{

    self = [super init];

    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    long lrUserBlocked = [[userProfile objectForKey:@"IsDeleted"] integerValue];

    if([[LoginRadiusSDK sharedInstance] useKeychain])
    {
        //save to keychain
        NSString *appIdentifierPrefix = [self bundleSeedID];
        
        NSString *groupKey = [NSString stringWithFormat:@"%@.%@",appIdentifierPrefix,[LoginRadiusSDK siteName]];

        A0SimpleKeychain *keychain = [A0SimpleKeychain keychainWithService:[LoginRadiusSDK siteName] accessGroup:groupKey];
        
        [keychain setString:token forKey:@"lrAccessToken"];
        
        BOOL yes = YES;
        BOOL no = NO;
        
        if(lrUserBlocked == 0) {
            [keychain setData: [NSData dataWithBytes:&(yes) length:sizeof(BOOL)] forKey:@"isLoggedIn" ];
            [keychain setData:[NSData dataWithBytes:&no length:sizeof(BOOL)] forKey:@"lrUserBlocked"];
        } else
        {
            [keychain setData:[NSData dataWithBytes:&yes length:sizeof(BOOL)] forKey:@"lrUserBlocked"];
        }
        
        NSMutableDictionary* profile;
        if (![userProfile objectForKey:@"errorCode"]) {
            profile = [[userProfile mutableCopy] replaceNullWithBlank];
            NSData *profileData = [NSKeyedArchiver archivedDataWithRootObject:profile];
            [keychain setData:profileData forKey:@"lrUserProfile"];
        }
    }else
    {
        //save to user defaults
        [defaults setObject:token forKey:@"lrAccessToken"];
        //User is not blocked
        if(lrUserBlocked == 0) {
            [defaults setInteger:true forKey:@"isLoggedIn"];
            [defaults setInteger:false forKey:@"lrUserBlocked"];
        } else {
            [defaults setInteger:true forKey:@"lrUserBlocked"];
        }
        
        NSMutableDictionary* profile;
        if (![userProfile objectForKey:@"errorCode"]) {
            profile = [[userProfile mutableCopy] replaceNullWithBlank];
            [defaults setObject:profile forKey:@"lrUserProfile"];
        }

    }
    
    return self;
}

- (BOOL) logout
{
    //if its already logged out then don't do anything
    //and if user specify don't invalidate the access token

    if(![self isLoggedIn])
    {
        return NO;
    }

    [[AuthenticationAPI authInstance] invalidateAccessToken: [self accessToken] completionHandler:^(NSDictionary * _Nullable data, NSError * _Nullable error)
    {
        if (error)
        {
            NSLog(@"Failed to invalidate LRtoken: %@",[error localizedDescription]);
        }else{
#ifdef DEBUG
            NSLog(@"Succesfully invalidate LRToken");
#endif
        }
    }];
    
    if([[LoginRadiusSDK sharedInstance] useKeychain])
    {
        NSString *appIdentifierPrefix = [self bundleSeedID];
        
        NSString *groupKey = [NSString stringWithFormat:@"%@.%@",appIdentifierPrefix,[LoginRadiusSDK siteName]];

        A0SimpleKeychain *keychain = [A0SimpleKeychain keychainWithService:[LoginRadiusSDK siteName] accessGroup:groupKey];
        
        [keychain deleteEntryForKey:@"isLoggedIn"];
        [keychain deleteEntryForKey:@"lrAccessToken"];
        [keychain deleteEntryForKey:@"lrUserBlocked"];
        [keychain deleteEntryForKey:@"lrUserProfile"];

    }else
    {
        NSUserDefaults *lrUserDefault = [NSUserDefaults standardUserDefaults];
        [lrUserDefault removeObjectForKey:@"isLoggedIn"];
        [lrUserDefault removeObjectForKey:@"lrAccessToken"];
        [lrUserDefault removeObjectForKey:@"lrUserBlocked"];
        [lrUserDefault removeObjectForKey:@"lrUserProfile"];
        [[NSUserDefaults standardUserDefaults] synchronize];
    }
    
    return YES;
}

- (BOOL) isLoggedIn
{
    return ([self accessToken] && [self userProfile]);
}

- (NSString *)bundleSeedID {
    NSDictionary *query = [NSDictionary dictionaryWithObjectsAndKeys:
                           (__bridge NSString *)kSecClassGenericPassword, (__bridge NSString *)kSecClass,
                           @"bundleSeedID", kSecAttrAccount,
                           @"", kSecAttrService,
                           (id)kCFBooleanTrue, kSecReturnAttributes,
                           nil];
    CFDictionaryRef result = nil;
    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, (CFTypeRef *)&result);
    if (status == errSecItemNotFound)
        status = SecItemAdd((__bridge CFDictionaryRef)query, (CFTypeRef *)&result);
    if (status != errSecSuccess)
        return nil;
    NSString *accessGroup = [(__bridge NSDictionary *)result objectForKey:(__bridge NSString *)kSecAttrAccessGroup];
    NSArray *components = [accessGroup componentsSeparatedByString:@"."];
    NSString *bundleSeedID = [[components objectEnumerator] nextObject];
    CFRelease(result);
    return bundleSeedID;
}

@end

File: Sources/LRSession.swift
//
//  LRSession.swift
//  LoginRadiusSwift SDK
//
//  Created by Megha Agarwal.
//  Copyright © 2023 LoginRadius Inc. All rights reserved.
//

import Foundation

public class LRSession {
    
    static let sharedInstance = LRSession()
    
    
    public init() {
        
    }
    
    public var accessToken: String? {
        var token: String?
        
        if LoginRadiusSDK.sharedInstance.useKeychain {
            let appIdentifierPrefix = bundleSeedID()
            let groupKey = "\(appIdentifierPrefix).\(LoginRadiusSDK.sharedInstance.siteName)"
            let keychain = LRKeychainWrapper(service: LoginRadiusSDK.siteName(), accessGroup: groupKey)
            
            if let tokenData = keychain.dataForKey(key: "lrAccessToken"){
                token = String(data: tokenData as Data, encoding: .utf8)
            }
        } else{
            let defaults = UserDefaults.standard
            token = defaults.string(forKey: "lrAccessToken")
        }
        
        return token
    }
    
    public var userProfile: [String: Any]? {
        var profile: [String: Any]?
        
        if LoginRadiusSDK.sharedInstance.useKeychain {
            let appIdentifierPrefix = bundleSeedID()
            let groupKey = "\(appIdentifierPrefix).\(LoginRadiusSDK.sharedInstance.siteName)"
            let keychain = LRKeychainWrapper(service: LoginRadiusSDK.siteName(), accessGroup: groupKey)
            if let profileData = keychain.dataForKey(key: "lrUserProfile"){
                profile = NSKeyedUnarchiver.unarchiveObject(with: profileData as Data) as? [String: Any]
            }
        } else {
            let defaults = UserDefaults.standard
            profile = defaults.dictionary(forKey: "lrUserProfile")
        }
        
        return profile
    }
    
    
    
    public init(token: String, userProfile: [String: Any]) {
        
        let defaults = UserDefaults.standard
        let lrUserBlocked = userProfile["IsDeleted"] as? Int ?? 0
        
        if LoginRadiusSDK.useKeychain() {
            // save to keychain
            let appIdentifierPrefix = bundleSeedID()
            let groupKey = "\(appIdentifierPrefix).\(LoginRadiusSDK.sharedInstance.siteName)"
            let keychain = LRKeychainWrapper(service: LoginRadiusSDK.siteName(), accessGroup: groupKey)
            
            if let tokenData = token.data(using: .utf8) {
                keychain.setData(data: tokenData as NSData, forKey: "lrAccessToken")
            }
            if lrUserBlocked == 0 {
                var yes = true
                var no = false
                keychain.setData(data: Data(bytes: &yes, count: MemoryLayout<Bool>.size) as NSData, forKey: "isLoggedIn")
                keychain.setData(data: Data(bytes: &no, count: MemoryLayout<Bool>.size) as NSData, forKey: "lrUserBlocked")
            } else {
                var yes = true
                keychain.setData(data: Data(bytes: &yes, count: MemoryLayout<Bool>.size) as NSData, forKey: "lrUserBlocked")
            }
            var profile = userProfile
            if profile["errorCode"] == nil {
                profile = profile.replaceNullWithBlank()
                if let profileData = try? NSKeyedArchiver.archivedData(withRootObject: profile, requiringSecureCoding: false) {
                    keychain.setData(data: profileData as NSData, forKey: "lrUserProfile")
                }
            }
        }else
        {
            defaults.set(token, forKey: "lrAccessToken")
            
            if lrUserBlocked == 0 {
                defaults.set(true, forKey: "isLoggedIn")
                defaults.set(false, forKey: "lrUserBlocked")
            } else {
                defaults.set(true, forKey: "lrUserBlocked")
            }
            
            var profile = userProfile
            if profile["errorCode"] == nil {
                profile = profile.replaceNullWithBlank()
                defaults.set(profile, forKey: "lrUserProfile")
            }
        }
        
        
    }
    
    
    
    public func logout() -> Bool {
        // if it's already logged out then don't do anything
        // and if the user specifies don't invalidate the access token
        
        if !isLoggedIn() {
            return false
        }
        
        AuthenticationAPI.shared.invalidateAccessToken(accessToken!) { (data, error) in
            if let error = error {
                NSLog("Failed to invalidate LRtoken: \(error.localizedDescription)")
            } else {
#if DEBUG
                NSLog("Successfully invalidate LRToken")
#endif
            }
        }
        
        if LoginRadiusSDK.useKeychain() {
            let appIdentifierPrefix = bundleSeedID()
            let groupKey = "\(appIdentifierPrefix).\(LoginRadiusSDK.siteName())"
            let keychain = LRKeychainWrapper(service: LoginRadiusSDK.siteName(), accessGroup: groupKey)
            
            keychain.deleteEntryForKey(key: "isLoggedIn")
            keychain.deleteEntryForKey(key: "lrAccessToken")
            keychain.deleteEntryForKey(key: "lrUserBlocked")
            keychain.deleteEntryForKey(key: "lrUserProfile")
        } else {
            let lrUserDefault = UserDefaults.standard
            lrUserDefault.removeObject(forKey: "isLoggedIn")
            lrUserDefault.removeObject(forKey: "lrAccessToken")
            lrUserDefault.removeObject(forKey: "lrUserBlocked")
            lrUserDefault.removeObject(forKey: "lrUserProfile")
            UserDefaults.standard.synchronize()
        }
        
        return true
    }
    
    
    public func isLoggedIn() -> Bool {
        var result:Bool = (accessToken != nil && userProfile != nil)
        return result
    }
    
    public func bundleSeedID() -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: "bundleSeedID",
            kSecAttrService as String: "",
            kSecReturnAttributes as String: kCFBooleanTrue as Any
        ]
        
        var result: AnyObject?
        var status = SecItemCopyMatching(query as CFDictionary, &result)
        
        if status == errSecItemNotFound {
            status = SecItemAdd(query as CFDictionary, &result)
        }
        
        if status != errSecSuccess {
            return nil
        }
        
        if let accessGroup = (result as? [String: Any])?[kSecAttrAccessGroup as String] as? String,
           let bundleSeedID = accessGroup.components(separatedBy: ".").first {
            return bundleSeedID
        }
        
        return nil
    }
}
